### PR TITLE
In 867 financial response API tests

### PR DESCRIPTION
**Pull request by jamesrwarren**
## Purpose

API tests for finance functionality

## Approach

As usual, tried to get the most existing statuses as possible. I've included the SQL for this as it may be very welcome if we have to find new test cases!

Basic rules I've followed. Where it's any non lookup field then a couple of examples of null and a couple with values is fine. Where it's a lookup field then one of each type in the data set is what we've gone for.

We have got lookups for every migrated field as well as some fields that are generated based on migrated fields (status, sop status etc).

Also after it was mentioned about errors not being obvious I realised that logging with an error code doesn't create a non zero exit code so made the errors a bit more visible.

Lastly, wanted to be able to handle nested objects better in responses so created a recursive function to handle any number of nested arrays in a json object.

## Learning

NA
## Checklist

* [x] I have performed a self-review of my own code
* [ ] I have done an adhoc run against preprod (only needed for high complexity PRs)
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation where relevant
* [x] I have added tests to prove my work
* [ ] The product team have tested these changes

### CODE REVIEW

File: migration_steps/validation/api_tests/shared/api_test.py [hunk 1]
```diff
@@ -310,28 +311,48 @@ def get_formatted_api_response(
...
+                            response_struct[json_locator] = (
+                                response_struct[json_locator] + result + "|"
+                            )
+                        except Exception:
+                            response_struct[json_locator] = result + "|"
 
         log.debug(f"returning: {response_struct}")
         return response_struct
 
+    def get_json_block_results(self, json_item_to_inspect, json_locator):
```
Comment (townxelliot) [inline migration_steps/validation/api_tests/shared/api_test.py:339]: My suggestion would be to use jsonpath to apply the selector to the response JSON. That way you've got a lot more flexibility and you can remove all the code for managing wildcards in the path and hand it over to the library.

File: scripts/api_tests_generator/preprod_cases/invoices.csv [hunk 2]
```diff
@@ -0,0 +1,69 @@
+endpoint,entity_ref,json_locator,test_purpose
```
Comment (townxelliot) [inline scripts/api_tests_generator/preprod_cases/invoices.csv:1]: Could these be converted to the same dot notation as the other tests (e.g. deputy_warnings.csv)?

**Review by townxelliot — DISMISSED** [review]
Looks good, made a suggestion to make all the JSON selectors consistent.

File: scripts/api_tests_generator/preprod_cases/invoices.csv (hunk 2 above)
Comment (jamesrwarren) [reply scripts/api_tests_generator/preprod_cases/invoices.csv:1]: Yeah good call. It's so permissive I hadn't noticed I'd missed some so good catch as it still generated the files fine :)

**Pull request merged**